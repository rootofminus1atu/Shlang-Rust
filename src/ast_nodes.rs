use crate::spans::*;
use std::collections::*;
#[derive(Clone, Debug, PartialEq)]

pub enum Control {
    Return(Box<Value>, Type),
    Result(Box<Value>, Type),
    Break,
}
impl From<Control> for Value {
    fn from(x: Control) -> Self {
        Value::Control(x)
    }
}
pub trait Conv{
    fn to_node(&self, span: Span)->Node;
}
#[derive(Clone, Debug, PartialEq)]
pub enum Value {
    Null,
    Void,
    Control(Control),
    Num(f64),
    Bool(bool),
    Str(String),
    Function(Function),
    BuiltinFunc(BuiltinFunc),
}
impl Value {
    pub fn get_type(&self) -> Type {
        match *self {
            Value::Bool(_) => return Type::Bool,
            Value::BuiltinFunc(_) | Value::Function(_) => return Type::Function,
            Value::Null => return Type::Null,
            Value::Void => return Type::Void,
            Value::Str(_) => return Type::Str,
            Value::Num(_) => return Type::Num,
            Value::Control(_) => return Type::Never,
        }
    }
}
impl Conv for Value {
    fn to_node(&self, span: Span)->Node {
        Node::value(self.clone(), span)
    }
}
#[derive(Clone, Debug, PartialEq)]
pub enum Type {
    Null,
    Void,
    Num,
    Bool,
    Str,
    Function,
    Never,
}
impl Type {
    pub fn is_void(&self) -> bool {
        self == &Self::Void
    }
    pub fn is_numeric(&self) -> bool {
        match self {
            Self::Bool | Self::Num => return true,
            _ => return false,
        }
    }
}
pub type TypedValue = (Value, Type);
#[derive(Clone, Debug, PartialEq)]
pub struct Function {
    block: Box<Block>,
    args: Vec<String>,
}
type ValueStream = Vec<Value>;
#[derive(Clone, Debug, PartialEq)]
pub struct BuiltinFunc {
    function: fn(ValueStream) -> Value,
    arg_size: i16,
}

#[derive(Clone, Debug, PartialEq)]
pub enum Node {
    Value(Spanned<Box<Value>>),
    Block(Spanned<Block>),
    DoBlock(Spanned<DoBlock>),
    BinaryNode(Spanned<BinaryNode>),
    UnaryNode(Spanned<UnaryNode>),
    ResultNode(Spanned<Box<Node>>),
    ReturnNode(Spanned<Box<Node>>),
    BreakNode(Span),
    Declaration(Spanned<Declaration>),
    Assignment(Spanned<Assignment>),
    Variable(Spanned<Variable>),
    Call(Spanned<Call>),
    Branch(Spanned<Branch>),
    Loop(Spanned<Loop>),
    While(Spanned<While>),
}
 
impl Node {
    pub fn block(body: NodeStream, span: Span) -> Self {
        return Node::Block(Spanned {
            unspanned: Block {
                body: Box::new(body),
            },
            span,
        });
    }
    pub fn value(value: Value, span: Span) -> Self {
        Node::Value(Spanned {
            unspanned: Box::new(value),
            span,
        })
    }
}

macro_rules! nodes_from {
    ($($name:ident)*) => {
        $(
            impl Conv for $name {
                fn to_node(&self,span:Span) -> Node {
                    Node::$name(Spanned {
                        unspanned: self.clone(),
                        span
                    })
                }
            }
        )*
    }
}
nodes_from! { UnaryNode DoBlock BinaryNode Call Variable Assignment Declaration Branch While Loop Block}
pub type NodeStream = Vec<Node>;
pub type NodeRef = Box<Node>;

#[derive(Clone, Debug, PartialEq)]
pub struct Block {
    pub body: Box<NodeStream>,
}
pub type BlockRef = Box<Block>;
#[derive(Clone, Debug, PartialEq)]
pub enum BinaryOp {
    ADD,
    SUBTRACT,
    DIVIDE,
    MULTIPLY,
    MODULO,
    AND,
    OR,
    ISEQUAL,
    ISDIFERENT,
    GREATER,
    LESSER,
    GREATER_EQUAL,
    LESSER_EQUAL,
}
#[derive(Clone, Debug, PartialEq)]
pub struct BinaryNode {
    pub kind: BinaryOp,
    pub left: NodeRef,
    pub right: NodeRef,
}
#[derive(Clone, Debug, PartialEq)]
pub enum UnaryOp {
    NEGATIVE,
    NOT,
}
#[derive(Clone, Debug, PartialEq)]
pub struct UnaryNode {
    pub kind: UnaryOp,
    pub object: NodeRef,
}
#[derive(Clone, Debug, PartialEq)]
pub struct Declaration {
    pub var_name: String,
    pub value: NodeRef,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Assignment {
    pub var_name: String,
    pub value: NodeRef,
}
#[derive(Clone, Debug, PartialEq)]
pub struct Variable {
    pub name: String,
}
#[derive(Clone, Debug, PartialEq)]
pub struct Call {
    pub callee: NodeRef,
    pub args: Box<NodeStream>,
}
#[derive(Clone, Debug, PartialEq)]
pub struct DoBlock {
    pub body: Box<NodeStream>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Branch {
    pub condition: NodeRef,
    pub if_block: BlockRef,
    pub else_block: Option<BlockRef>,
}
#[derive(Clone, Debug, PartialEq)]
pub struct Loop {
    pub proc: BlockRef,
}
#[derive(Clone, Debug, PartialEq)]
pub struct While {
    pub condition: NodeRef,
    pub proc: BlockRef,
}
#[derive(Clone, Debug, PartialEq)]
pub struct Scope {
    pub parent: Option<Box<Scope>>,
    pub var_map: HashMap<String, Value>,
}

impl Scope {
    pub fn get_var(&self, var_name: String) -> Option<Value> {
        if let Some(var) = self.var_map.get(&var_name) {
            return Some(var.clone());
        }
        if let Some(parent) = &self.parent {
            return parent.get_var(var_name);
        }

        None
    }
    pub fn define(&mut self, var_name: String, val: Value) {
        dbg!(self.var_map.insert(var_name, val));
        dbg!(&self.var_map);
    }
    pub fn new(parent: Option<Box<Scope>>, var_map: HashMap<String, Value>) -> Self {
        Scope { parent, var_map }
    }
    pub fn travel(&mut self) {
        let Some(parent) = self.parent.clone() else {panic!()};
        let grandpa = parent.clone().parent;
        self.parent = grandpa;
        self.var_map = parent.var_map;
    }
    pub fn assign(&mut self, var_name: String, value: Value) -> Option<Value> {
        if let Some(var) = self.var_map.get_mut(&var_name) {
            *var = value;
            return Some(var.clone());
        }
        if let Some(parent) = &mut self.parent {
            return parent.assign(var_name, value);
        }
        None
    }
}
