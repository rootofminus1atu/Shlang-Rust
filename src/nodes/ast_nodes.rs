use std::collections::*;

use crate::spans::*;

pub struct AST {
    functions: HashMap<String, Function>,
    variables: HashMap<String, Value>,
    structs: HashMap<String, Struct>,
}
#[derive(Clone, Debug, PartialEq)]
pub enum Value {
    Null,
    Void,
    Num(f64),
    Bool(bool),
    Str(String),
    Function(Function),
    BuiltinFunc(BuiltinFunc),
    Struct(Struct),
    StructRef(u32),
}
pub type TypedValue = (Value, Type);
pub type ValueStream = Vec<Value>;
impl Value {
    pub fn get_type(&self) -> Type {
        match self {
            Value::Bool(_) => Type::Bool,
            Value::BuiltinFunc(_) | Value::Function(_) => Type::Function,
            Value::Null => Type::Null,
            Value::Void => Type::Void,
            Value::Str(_) => Type::Str,
            Value::Num(_) => Type::Num,

            Value::Struct(s) => Type::UserDefined(s.id.clone()),
            Value::StructRef(id) => Type::Ref(*id),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct Struct {
    pub id: String,
    pub env: Scope,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Function {
    pub block: Block,
    pub args: Vec<String>,
}
impl Function {
    pub fn new(block: Block, args: Vec<String>) -> Self {
        Self { block, args }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct BuiltinFunc {
    pub function: fn(HashMap<String, Value>, Vec<Value>) -> Value,
    pub arg_size: i16,
}
impl From<Function> for Value {
    fn from(x: Function) -> Self {
        Value::Function(x)
    }
}
#[derive(Clone, Debug, PartialEq)]
pub enum Type {
    Null,
    Void,
    Num,
    Bool,
    Str,
    Function,

    Ref(u32),
    UserDefined(String),
}
impl Type {
    pub fn is_void(&self) -> bool {
        self == &Self::Void
    }
    pub fn is_numeric(&self) -> bool {
        matches!(self, Self::Bool | Self::Num)
    }
}
pub type Block = Vec<Spanned<Expr>>;
#[derive(Clone, Debug, PartialEq)]
pub enum Expr {
    Statement(Statement),
    Value(Value),
    BinaryNode(BinaryOp, Spanned<Box<Expr>>, Spanned<Box<Expr>>),
    UnaryNode(UnaryOp, Spanned<Box<Expr>>),
    Variable(String),
    Call(Spanned<Box<Expr>>, Vec<Spanned<Expr>>),
    Branch(Spanned<Box<Expr>>, Block, Option<Block>),
    Loop(Block),
    While(Spanned<Box<Expr>>, Block),
    DoBlock(Block),
    Constructor(String, HashMap<String, Spanned<Expr>>),
    StructDef(String, Vec<Spanned<Field>>),
    FieldAccess {
        target: Spanned<Box<Expr>>,
        requested: Spanned<Box<Expr>>,
    },
}
impl Expr {
    pub fn to_spanned(self, span: Span) -> Spanned<Self> {
        Spanned::new(self, span)
    }
}
#[derive(Clone, Debug, PartialEq)]
pub enum Statement {
    Return(Spanned<Box<Expr>>),
    Break,
    Continue,
    Declaration(String, Spanned<Box<Expr>>),
    Assignment(Spanned<Box<Expr>>, Spanned<Box<Expr>>),
}
impl Statement {
    pub fn to_exprspan(self, span: Span) -> Spanned<Expr> {
        Spanned::new(Expr::Statement(self), span)
    }
}
impl From<Statement> for Expr {
    fn from(value: Statement) -> Self {
        return Expr::Statement(value);
    }
}
#[derive(Clone, Debug, PartialEq)]
pub enum BinaryOp {
    ADD,
    SUBTRACT,
    DIVIDE,
    MULTIPLY,
    MODULO,
    AND,
    OR,
    ISEQUAL,
    ISDIFERENT,
    GREATER,
    LESSER,
    GREATER_EQUAL,
    LESSER_EQUAL,
}

#[derive(Clone, Debug, PartialEq)]
pub enum UnaryOp {
    NEGATIVE,
    NOT,
}
#[derive(Clone, Debug, PartialEq)]
pub enum Field {
    Declaration(String, Spanned<Expr>),
    StructDef(String, Vec<Spanned<Field>>),
}

pub type VarMap = HashMap<String, Value>;
#[derive(Clone, Debug, PartialEq)]
pub struct Scope {
    pub parent: Option<Box<Scope>>,
    pub vars: VarMap,
    pub structs: HashMap<String, Struct>,
}

impl Scope {
    pub fn get_var(&self, var_name: &String) -> Option<Value> {
        if let Some(var) = self.vars.get(var_name) {
            return Some(var.clone());
        }
        if let Some(parent) = &self.parent {
            return parent.get_var(var_name);
        }
        None
    }
    pub fn get_struct(&self, struct_name: &String) -> Option<Struct> {
        if let Some(obj) = self.structs.get(struct_name) {
            return Some(obj.clone());
        }
        if let Some(parent) = &self.parent {
            return parent.get_struct(struct_name);
        }
        None
    }
    pub fn define(&mut self, var_name: String, val: Value) {
        if let Value::Struct(obj) = &val {
            self.structs.insert(var_name.clone(), obj.clone());
        }
        self.vars.insert(var_name, val);
    }
    pub fn new(parent: Option<Box<Scope>>, vars: VarMap, structs: HashMap<String, Struct>) -> Self {
        Scope {
            parent,
            vars,
            structs,
        }
    }
    pub fn new_child_in(parent: Scope) -> Self {
        Scope {
            parent: Some(Box::new(parent)),
            vars: HashMap::from([]),
            structs: HashMap::from([]),
        }
    }
    pub fn assign(&mut self, var_name: String, value: Value) -> Option<Value> {
        if let Some(var) = self.vars.get_mut(&var_name) {
            *var = value;
            return Some(var.clone());
        }
        if let Some(parent) = &mut self.parent {
            return parent.assign(var_name, value);
        }
        None
    }
}
