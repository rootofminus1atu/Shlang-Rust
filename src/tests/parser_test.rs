use crate::ast_nodes;
use crate::Parser;
use insta::*;
use std::io;
use std::io::Write;
use std::*;
use ast_nodes::*;
fn parse_full(source: &str) -> Block {
    let mut parser = Parser::new(source);
    return parser.batch_parse();
}
fn parse_expr(source: &str) -> Node {
    let mut parser = Parser::new(source);
    return parser.parse_expr();
}
macro_rules! test_func {
    ($func:expr,($($name:ident : $input:expr)*)) => {
        $(
        #[test]
        fn $name () {
            // Perform your desired operation using the input parameter
            assert_debug_snapshot!($func($input));
        }
        )*
    };
}
test_func!(parse_full,(
declare_and_assign_expression:"var a = 1+2+b;"
single_var_decl:"var a;"
multi_var_decl:"var a; var a;"
var_declare_and_assign_value:"var a = 2;"
));
test_func!(parse_expr,(
parse_nested_paren:"((1+2)+(1+2));"
parse_paren:"(1+2+b);"
expression_operators:"1*2+2-2/2%5 and 10 or 10!=1==2<4<=3>0>=3;"
unary_operator:"!true;"
multiple_unary:"!true + !true;"
unary_operator_paren:"!(0);"
unary_operators:"!0 +(-0);"
empty_call:"a();"
call:"a((1+2),true,\"1\");"
call_as_val:"a(1,1,1)+1;"
call_with_paren:"a((1),1,1);"
assignement:"b = 1;"
));
