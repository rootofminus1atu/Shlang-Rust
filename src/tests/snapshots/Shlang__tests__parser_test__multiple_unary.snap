---
source: src/tests/parser_test.rs
expression: "parse_expr(\"!true + !true;\")"
---
BinaryNode(
    Spanned {
        unspanned: BinaryNode {
            kind: ADD,
            left: UnaryNode(
                Spanned {
                    unspanned: UnaryNode {
                        kind: NOT,
                        object: Value(
                            Spanned {
                                unspanned: Bool(
                                    true,
                                ),
                                span: (
                                    1,
                                    5,
                                ),
                            },
                        ),
                    },
                    span: (
                        1,
                        5,
                    ),
                },
            ),
            right: UnaryNode(
                Spanned {
                    unspanned: UnaryNode {
                        kind: NOT,
                        object: Value(
                            Spanned {
                                unspanned: Bool(
                                    true,
                                ),
                                span: (
                                    9,
                                    13,
                                ),
                            },
                        ),
                    },
                    span: (
                        9,
                        13,
                    ),
                },
            ),
        },
        span: (
            6,
            14,
        ),
    },
)
